File: Packages/NotebarCore/Sources/NotebarCore/Models/LinkTombstone.swift
import Foundation

/// The pure "is this chip's target gone" predicate behind spec §6.4
/// deliverable 3's tombstones — split out from the AppKit styling that
/// actually paints a tombstone (`NoteChipStyling` in the app target, which
/// this package must never import — spec section 3, rule 1) so the decision
/// itself is unit-testable here, the same split `NoteListMarkers`/
/// `NoteListEditing` already draw between "the pure logic" and "the
/// `NSTextView` glue."
///
/// **Why this can't be answered from the `link` table.** `LinkSchema`'s
/// `AFTER DELETE` triggers cascade: deleting a note or task also deletes
/// every `link` row that touches it, on either end. So by the time a chip
/// needs restyling, the row behind it is already gone too — a tombstone
/// can't be detected by asking "does a link row still point here," only by
/// asking "does the target itself still exist." Callers answer that once,
/// not per chip — see `PanelViewModel.existingLinkTargets()`.
public enum LinkTombstone {
    /// `nil` when `url` isn't a chip this app ever wrote at all (a foreign
    /// scheme, or a malformed `notebar://` URL) — mirrors `LinkURL.parse`'s
    /// own `nil` case, so a caller can tell "not a chip" apart from "a chip
    /// whose target is gone." `true` when it is a chip and `existingTargets`
    /// (every note/task id known to still exist, batched ahead of time) does
    /// not contain its target.
    public static func isTombstone(url: URL, existingTargets: Set<LinkTarget>) -> Bool? {
        guard let target = LinkURL.parse(url) else { return nil }
        return !existingTargets.contains(target)
    }
}

File: windows/Notebar.Core/Models/LinkTombstone.cs
namespace Notebar.Core.Models;

/// <summary>Whether a chip's target no longer exists.</summary>
/// <remarks>
/// This cannot be answered from the link table. Deleting a note or task cascades
/// and removes every link row that touches it, on either end — so by the time a
/// chip needs restyling, the row behind it is gone too. A tombstone is detectable
/// only by asking "does the target itself still exist," which callers answer once
/// for the whole document rather than once per chip.
///
/// Split out from the styling that actually paints a tombstone so the decision
/// is unit-testable without a WebView.
/// </remarks>
public static class LinkTombstone
{
    /// <summary>Null when <paramref name="url"/> is not a chip this app ever wrote
    /// — a foreign scheme, or a malformed notebar:// URL — mirroring
    /// <see cref="LinkUrl.Parse"/>, so a caller can tell "not a chip" apart from
    /// "a chip whose target is gone". True when it is a chip and
    /// <paramref name="existingTargets"/> does not contain its target.</summary>
    public static bool? IsTombstone(string url, IReadOnlySet<LinkTarget> existingTargets)
    {
        var target = LinkUrl.Parse(url);
        if (target is null) return null;
        return !existingTargets.Contains(target.Value);
    }
}
